### ISSUE TITLE
k8s operator adds suffixes to `tailscale.com/hostname`

### ISSUE THREAD

**Issue by aviau** [tailscale/tailscale#10604]
### What is the issue?

When using the k8s operator, I set `tailscale.com/hostname` annotations to ensure that the proxies use a specific hostname.

Instead of respecting the value that I set in `tailscale.com/hostname`, the operator will add the `-1` suffix if it already exists.

This isn's acceptable for my use case where the hostname matters. In these cases, I think the k8s operator should "take over" the existing device, probably by deleting it and creating a new one.

The same problem happens not only for the proxies but for the operator itself:

<img width="1139" alt="Screenshot 2023-12-14 at 3 39 09 PM" src="https://github.com/tailscale/tailscale/assets/2706882/ed8df71c-701a-4f23-8d99-86c98ad72af7">

**Comment by knyar** [comment]
I believe the behaviour you are seeing is not specific to k8s-operator. As mentioned in the [machine names](https://tailscale.com/kb/1098/machine-names) doc, when a machine registers with a hostname that already exists, it gets a numeric suffix to allow users to disambiguate two devices. You can rename devices in Admin Console, or upvote [this feature request](https://github.com/tailscale/tailscale/issues/4121) if you would like to see an API endpoint for changing the name.

The operator currently does keep track of Tailscale nodes it created and deletes them [as part of cleanup](https://github.com/tailscale/tailscale/blob/dc2792aaee2b6cee72fed1e0c597f230b934d730/cmd/k8s-operator/sts.go#L134). So duplicate names are likely to be real conflicts of intent rather than artefacts left by the operator as part of its normal operation.

We should add a note to the operator docs to clarify that the annotation sets the hostname reported by the new node, but there is no guarantee that it will be used verbatim.

Perhaps we could also find a way to surface to the user that the actual hostname is different to the one they specified in the annotation. I'd suggest we keep this issue as a feature request for that.

**Comment by aviau** [comment]
@knyar 

> when a machine registers with a hostname that already exists, it gets a numeric suffix to allow users to disambiguate two devices

I can see this leading to a lot of issues. In the case of nodes that need stable names (not laptops, but web services for example), getting a new name means nobody can trust the tailscale DNS and have to revert to another way of resolving the tailscale IP for that service? What are the recommended best practices there?

**Comment by passcod** [comment]
Second this, it's weird and annoying that whenever the operator pod gets restarted, it changes name to `{hostname}-1`. Combined with #10102 it means a proxy-on operator is very fragile, a node going down will almost always lead to needing manual intervention in the control panel and every application that relies on the tailscale-mediated kube api connection will break.

**Comment by knyar** [comment]
> whenever the operator pod gets restarted, it changes name to {hostname}-1

This should not be happening unless you have multiple operators attempting to use the same hostname, or if the operator cannot write its state to the [k8s secret](https://github.com/tailscale/tailscale/blob/36efc508176b79e035332b95d414048f6ed46fa7/cmd/k8s-operator/operator.go#L106). Could you please share more details about your usage of the operator? (version, the way you installed it, and current configuration).

**Comment by passcod** [comment]
Yeah, I was about to edit. I'm filing a new issue and reading the logfile for any necessary redaction noticed it was trying (and failing, due to the issue) to write to a secret.

**Comment by mickvangelderen** [comment]
Given that this hasn't been implemented yet, what is a good way to remedy a situation where:

- the tailscale operator was down
- finalizers couldn't run
- cluster was recreated
- now machines are named x-1

I suppose the old machines should be deleted (how to best do this) and somehow the tailscale operator needs to recreate the machines (how to do this) and then the x-1 names need to be deleted.
